src/util/modules.py

```diff
@@ -549,20 +549,23 @@ def force_install_module(cr, module, if_installed=None, reason="it has been expl
                                            already installed
     :return str: the *new* state of the module
     """
-    try:
-        return _force_install_module(cr, module, if_installed, reason)
-    except UnknownModuleError:
-        if version_gte("saas~14.5"):
-            # We must delay until the modules actually exists. They are added by the auto discovery process.
+    if version_gte("saas~14.5"):
+        if not if_installed or modules_installed(cr, *if_installed):
```

Comment:
We have a similar issue here. The modules in `if_installed` may not be discovered yet. Continuing the same discussion I referred before (with CHS), I think we can switch `__modules_auto_discovery_force_installs` into a dict: `module: if_installs`. And do the check post base load. A strict check, such that if a module in if_installed does not exist we get an error in the logs. This will detect issues in CI at the source.

Reply:
Indeed we can leave it as is. In any case current code is the product of the limitations you are fixing here. Thus some alternative solutions were found (like manually calling deps diffs) that would be obsolete afterwards. We can review those later.

Reply:
Note that if_installed check could check the actual `installed` state, vs the `to install/to upgrade` transient states.

Reply:
Given (our) low usage of the `if_installed` feature, a hard-fail in this case is OK.

Reply:
Added the validation `_assert_modules_exists`, but kept all installed states for not. I think `to upgrade` is also valid for custom modules if the upgrade is jumping multiple versions and the check is not done in the first upgrade step so the `installed` is already `to upgrade` in pre- base